common/cuda_hip/components/bitvector_kernels.cpp

```diff
@@ -0,0 +1,55 @@
+// SPDX-FileCopyrightText: 2017 - 2025 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "core/components/bitvector_kernels.hpp"
+
+#include <thrust/fill.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/scan.h>
+
+#include <ginkgo/core/base/intrinsics.hpp>
+
+#include "common/cuda_hip/base/thrust.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace GKO_DEVICE_NAMESPACE {
+namespace bitvector {
+
+
+template <typename IndexType>
+void compute_bits_and_ranks(
+    std::shared_ptr<const DefaultExecutor> exec, const IndexType* indices,
+    IndexType num_indices, IndexType size,
+    typename device_bitvector<IndexType>::storage_type* bits, IndexType* ranks)
+{
+    const auto policy = thrust_policy(exec);
+    using bv = device_bitvector<IndexType>;
+    using storage_type = typename bv::storage_type;
+    const auto num_blocks = ceildiv(size, bv::block_size);
+    thrust::fill_n(policy, bits, num_blocks, 0u);
+    thrust::for_each_n(
+        policy, indices, num_indices, [bits] __device__(IndexType idx) {
+            constexpr auto block_size = device_bitvector<IndexType>::block_size;
+            const auto block = idx / block_size;
+            const auto local = idx % block_size;
+            atomicOr(bits + block, storage_type{1} << local);
```

Comment:
We can atomicOr in the shared memory rather than the global memory.
The global write will not need the initialization and the atomic on the global memory